### PR TITLE
Add ZIP TDG configs support

### LINKED ISSUE #293: [8pt] CI: Add tests with TDG

### LINKED ISSUE #327: [4pt] Support uploading TDG config as a ZIP archive

### PR DESCRIPTION
**Pull request by hackallcode**
Closes #293
Closes #327

Before the patch, it was possible to upload TDG config only in folder format. Also, there were no TDG tests in CI.

Changes:
- Now it's possible to upload the TDG config in ZIP format;
- Added tests for TDG (including `config_upload_tdg` molecule test);
- `config_upload` test fixed and added to CI matrix.

### CODE REVIEW

File: doc/app_config.md [hunk 1]
```diff
@@ -93,11 +93,10 @@ cartridge_app_config_upload_mode: 'tdg'
...
 To upload file config, you can use:
 - Lua mode (default for YAML files);
-- HTTP mode (default for not-YAML files).
+- HTTP mode (default for not-YAML files);
+- TDG mode (available only for ZIP files).
 
-To upload directory config, you can use:
-- HTTP mode (default);
-- TDG mode.
+To upload directory config, you can use TDG mode only.
```
Comment (dokshina) [inline doc/app_config.md]: You can upload the config directory only via TDG mode.

File: library/cartridge_apply_app_config.py [hunk 2]
```diff
@@ -79,23 +86,65 @@ def apply_tdg_config(console_sock, path):
...
+def pack_zip(path):
+    zip_config_path = os.path.abspath(os.path.join(path, os.pardir, 'config-%s.zip' % str(uuid.uuid4())))
 
-        with zipfile.ZipFile(zip_config_path, 'w') as zf:
-            for root, _, file_paths in os.walk(config_path):
-                rel_dir_path = os.path.relpath(root, config_path)
+    with zipfile.ZipFile(zip_config_path, 'w') as zf:
+        if not os.path.isdir(path):
+            base_name = os.path.basename(path)
+            zf.write(path, base_name)
```
Comment (dokshina) [inline library/cartridge_apply_app_config.py]: It's for uploading file config? Please, add a comment.

File: molecule/config_upload_tdg/converge.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
+---
+
+- name: 'Test file config'
+  hosts: cluster
+  become: true
+  become_user: root
+  gather_facts: false
+  vars:
+    cartridge_app_config_upload_mode: 'tdg'
+    cartridge_eval_body: >-
```
Comment (dokshina) [inline molecule/config_upload_tdg/converge.yml:10]: Looks strange without `eval` step. Let's place these vars on each import separately.

Comment (hackallcode) [reply molecule/config_upload_tdg/converge.yml:10]: No. Eval is here.

File: library/cartridge_apply_app_config.py (hunk 2 above)
Comment (hackallcode) [reply library/cartridge_apply_app_config.py]: Yes, it is for file zipping, but it is not used. I removed this

File: .github/helpers/count_molecule_matrix.py [hunk 4]
```diff
@@ -57,13 +67,31 @@ def main(event_name, repo_owner, review_state, ref):
...
+
     print(f'::set-output name=ce-tests-found::{"true" if len(ce_matrix) > 0 else "false"}')
     print(f'::set-output name=ee-tests-found::{"true" if len(ee_matrix) > 0 else "false"}')
+    print(f'::set-output name=tdg-tests-found::{"true" if len(tdg_matrix) > 0 else "false"}')
 
     print(f'::set-output name=ce-matrix::{json.dumps({"include": ce_matrix})}')
     print(f'::set-output name=ee-matrix::{json.dumps({"include": ee_matrix})}')
+    print(f'::set-output name=tdg-matrix::{json.dumps({"include": tdg_matrix})}')
+
+    print('Final matrices:')
```
Comment (dokshina) [inline .github/helpers/count_molecule_matrix.py]: Maybe `Computed matrices`?

File: library/cartridge_apply_app_config.py [hunk 5]
```diff
@@ -104,14 +149,18 @@ def get_tdg_upload_mode(console_sock):
 
     control_console = helpers.get_control_console(console_sock)
     return control_console.eval_res_err('''
-        if admin ~= nil and admin.upload_config_api ~= nil then
+        function upload_config_api_exists()
+            return admin ~= nil and admin.upload_config_api ~= nil
```
Comment (dokshina) [inline library/cartridge_apply_app_config.py]: Maybe `if rawget(_G, 'admin) ~= nil`?

File: library/cartridge_apply_app_config.py [hunk 6]
```diff
@@ -165,7 +213,7 @@ def get_apply_config_func_for_lua(console_sock):
 
     control_console = helpers.get_control_console(console_sock)
 
-    return lambda path: patch_file_clusterwide(control_console, path)
+    return FILE_OUTPUT_FORMAT, lambda path: patch_file_clusterwide(control_console, path)
```
Comment (dokshina) [inline library/cartridge_apply_app_config.py:212]: `get_apply_config_func_for_lua` -> `get_output_format_and_apply_func_for_lua `?

File: library/cartridge_apply_app_config.py [hunk 7]
```diff
@@ -183,32 +231,32 @@ def get_apply_config_func_for_tdg(console_sock, upload_url, tdg_token):
     if tdg_upload_mode == 'http':
         assert upload_url is not None, 'Upload URL is required for TDG mode'
 
-        headers = get_tdg_auth_headers(console_sock, tdg_token)
+        headers = get_tdg_http_headers(console_sock, tdg_token)
 
-        return lambda path: send_on_http(upload_url, headers, path), ZIP_DIR_MODE
+        return ZIP_OUTPUT_FORMAT, lambda path: send_on_http(upload_url, headers, path)
```
Comment (dokshina) [inline library/cartridge_apply_app_config.py:232]: The same as above - reflet it in the function name.

File: library/cartridge_apply_app_config.py [hunk 8]
```diff
@@ -218,26 +266,28 @@ def apply_app_config(params):
...
     remote_config_path = os.path.join(params['remote_dir'], os.path.basename(params['local_config_path']))
 
-    upload_mode = get_upload_mode(
+    upload_mode = prepare_upload_mode(
         upload_mode=params['upload_mode'],
         remote_config_path=remote_config_path,
     )
 
-    apply_config, dir_mode = get_apply_config_func(
+    output_format, apply_config = get_apply_config_func(
```
Comment (dokshina) [inline library/cartridge_apply_app_config.py]: `apply_config_func`

**Review by dokshina — DISMISSED** [review]

**Review by dokshina — APPROVED** [review]

**Pull request merged**